File: C++/121. Best Time to Buy and Sell Stock.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    int maxProfit(std::vector<int>& prices) {
        int res = 0;
        if (prices.size() == 0)
            return res;
        int min = prices[0], max = prices[0];
        for (int i = 1; i < int(prices.size()); i++) {
            if (prices[i] < min) {
                min = prices[i];
                max = prices[i];
            } else if (prices[i] > max) {
                max = prices[i];
            }
            if (max - min > res)
                res = max - min;
        }
        return res;
    }

    /* another method
    int maxProfit(std::vector<int>& prices) {
        int res = 0;
        if (prices.size() == 0)
            return res;
        int min = prices[0];
        for (int i = 1; i < int(prices.size()); i++) {
            if (prices[i] < min)
                min = prices[i];
            if (prices[i] - min > res)
                res = prices[i] - min;
        }
        return res;
    }
    */
};

File: Go/121. Best Time to Buy and Sell Stock.go
func maxProfit(prices []int) int {
	res := 0
	if len(prices) == 0 {
		return res
	}
	
	min := prices[0]
	max := prices[0]
	for i := 0; i < len(prices); i++ {
		if prices[i] > max {
			max = prices[i]
		} else if prices[i] < min {
			min = prices[i]
			max = prices[i]
		}
		if max - min > res {
			res = max - min
		}
	}
	return res
}

/*
func maxProfit(prices []int) int {
	res := 0
	if len(prices) == 0 {
		return res
	}
	min := prices[0]
	for i := 0; i < len(prices); i++ {
		if prices[i] < min {
			min = prices[i]
		}
		if res < prices[i] - min {
			res = prices[i] - min
		}
	}
	return res
}
*/